### ISSUE TITLE
[Specification] VkImageFormatProperties::maxMipLevels does not restrict anything

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#670]
In the definition of `VkImageFormatProperties::maxMipLevels` it says:

> `maxMipLevels` is the maximum number of mipmap levels. `maxMipLevels` **must** either be equal to 1 (valid only if `tiling` is `VK_IMAGE_TILING_LINEAR`) or be equal to `⌈log2(max(width, height, depth))⌉ + 1`. `width`, `height`, and `depth` are taken from the corresponding members of `maxExtent`.

But `vkCreateImage` already has to have smaller extent than `maxExtent` and already has VU saying:

> `mipLevels` must be less than or equal to `⌊log2(max(extent.width, extent.height, extent.depth))⌋ + 1`

which makes it inevitably less than `maxMipLevels` .

So what purpose does `VkImageFormatProperties::maxMipLevels` serve? Only difference I can see is that it returns whether `LINEAR` format can have mip-maps, which could simply be served by `VkBool32`.

BTW: Why is one formula rounded up and the second rounded down??

**Comment by Tobski** [comment]
So this "maxMipLevels" keeps tripping us up too; unlike the rest of the queries which sort of "extend" capabilities beyond the usual numbers you get from vkGetPhysicalDeviceFormatProperties, this one is actually a restriction on those extended capabilities. Not confusing at all, right...? 😛 

Basically it's there as a "if you go to the higher (width/height/depth) dimensions advertised by this feature for a specific image, you might not be able to create `log2(max_dimension)` mip levels for that image - you are restricted to that lower number.

In other words, there are cases where you might be able to create a large image but not create a full mip stack for it.

> mipLevels must be less than or equal to VkImageFormatProperties::maxMipLevels (as
> returned by vkGetPhysicalDeviceImageFormatProperties with format, imageType, tiling,
> usage, and flags equal to those in this structure)

So this VU is correct, as maxMipLevels may be less than `log2(max_dimension)` in some specific circumstances. It's just not immediately obvious why that is...

**Comment by krOoze** [comment]
> [...] as maxMipLevels may be less than `log2(max_dimension)` in some specific circumstances.

Except it cannot:

> `maxMipLevels` is the maximum number of mipmap levels. `maxMipLevels` **must** be equal [!!!] to `⌈log2(max(width, height, depth))⌉ + 1`, where `width`, `height`, and `depth` are taken from the corresponding members of `maxExtent`, except when one of the following conditions is true, in which case it **may** instead be 1: [conditions]

therefore the VU you quote is redundant to the VU I quote.

`⌊log2(max(extent.width, extent.height, extent.depth))⌋ + 1` is strictly less than or equal to `⌈log2(max(width, height, depth))⌉ + 1`, because `(extent.width, extent.height, extent.depth)` must be less than or equal to `(width, height, depth)`.

**Comment by Tobski** [comment]
I remember looking at this and going "darn it, @krOoze is right - how did we manage that?". So that's... actually... useful? It makes this easier anyway 😅 

Fortunately, this redundancy got fixed a while ago with the whole image creation limits modifications, although that went the other way and removed the programmatic definition of the limit, instead just having it depend on maxMipLevels.

I think this can be closed now?
